### DISCUSSION TITLE
"Decoding" the input to make parameters (newbie question)

### DISCUSSION THREAD

**Discussion by fcorbelli** [junegunn/fzf#3379] [General]
I didn't know much of fzf, but a user of mine (https://github.com/fcorbelli/zpaqfranz/issues/63#issuecomment-1646869427) asked me to support it, so I'm going to ask a few questions that have probably been answered 1,000 times already  
Please be patient

Suppose we have some kind of encoded input-parameter.
For example
```
0003|test.txt
0004|test.txt
0002|/tmp/foo.txt
0003|/tmp/foo.txt
```

There is a number (a backup version) and a filename  
You usually get same filenames, BUT different versions

Suppose you want to run "something" (aka: the file extractor) selecting the 3nd line of the example by fzf
We need to take the version number (2 in this example) and the filename (/tmp/foo.txt)

```
zpaqfranz x thearchive.zpaq "/tmp/test.txt" -until 2 -to wherever
```

Is it possible and how?
I can, of course, change the output to make things easier (fixed length, comma delimited or whatever)

Thanks to all answers

**Comment by luckman212** [comment]
I believe the correct way to achieve this would be to add some shell script wrapper around it, something like

```bash
# set vars below as needed
ARCHIVE='/path/to/thearchive.zpaq'
DESTINATION='/path/to/destination/dir/'

while IFS='|' read -r VERSION FILENAME _ ; do
  zpaqfranz x "$ARCHIVE" "$FILENAME" -until $VERSION -to "$DESTINATION"
done < <(zpaqfranz fzf "$ARCHIVE" | fzf --multi --no-select-1)
```

**Reply by fcorbelli** [reply]
The more specific question is: so is it enough for me to make an output as indicated (version number | file name) and then it becomes the user's "problem" to use it somehow with fzf?

Another trivial question: can fzf sort the input, according to various choices, or does it really directly read stdin one line at a time (translation: I put the sorter inside zpaqfranz)?

Thank you very much!

**Reply by luckman212** [reply]
Yes, in my opinion the simple output with delimiter (either TAB `\t`, PIPE `|`, or NUL byte character) would be enough to make it interoperate nicely with fzf. The rest is up to the user how they want to handle it.

fzf does not sort the input, but as the user types to start filtering the file list, it will rank the results in order according to its matching algorithm. This can be adjusted via fzf flags such as `--algo`, `--tiebreak` (or completely disabled with `--no-sort`). The fzf manpage is extensive.

(**TL;DR**- no need to sort anything in zpaqfranz)

**Reply by fcorbelli** [reply]
I was referring to entering additional data to enable sorting, such as file size (this function exists for the Windows version with a kind of text GUI, instead of fzf)

```
0000999945|0001|file.txt
0000777124|0003|file.txt
0000777125|0002|file.txt
```

Anyway I will release a "quick-and-dirty" fzf-friendly output, I can always introduce improvements

**Reply by luckman212** [reply]
Ah, yes I understand. That might be useful for sure. You can control which columns are displayed in fzf using the `--with-nth` flag.

So in your example
```
0000999945|0001|file.txt
0000777124|0003|file.txt
0000777125|0002|file.txt
```

If you want you could feed this to fzf, sorting by the filesize (but keeping it hidden) with something like
```bash
#!/bin/bash

OUTPUT="0000999945|0001|file.txt
0000777124|0003|file.txt
0000777125|0002|file.txt"

echo "$OUTPUT" | sort -n -k1 | fzf --delimiter='\|' --with-nth=2,3
```
which will yield a file selector inside fzf like this
```
0003|file.txt
0002|file.txt
0001|file.txt
```
